### PR TITLE
fixes for virus scanner integration 

### PR DESCRIPTION
**Pull request by meal**
If virus scanner is disabled, then create Scan associated item with `clean` status

https://github.com/bitzesty/qae/issues/337 addresses it

### CODE REVIEW

File: app/models/audit_certificate.rb [hunk 1]
```diff
@@ -33,7 +33,11 @@ def reviewed?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
+      Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Line is too long. [106/100]<br>Redundant `self` detected.

File: app/models/audit_certificate.rb [hunk 2]
```diff
@@ -33,7 +33,11 @@ def reviewed?
...
+    else
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, audit_certificate_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Line is too long. [130/100]<br>Redundant `self` detected.

File: app/models/audit_certificate.rb [hunk 3]
```diff
@@ -33,7 +33,11 @@ def reviewed?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 4]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.file.url)
-    Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.file.url)
+      Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Line is too long. [105/100]<br>Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 5]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
...
+    else
+      Scan.create(filename: self.file.current_path, uuid: SecureRandom.uuid, form_answer_attachment_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Line is too long. [129/100]<br>Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 6]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.file.url)
-    Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.file.url)
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 7]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
+      Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Line is too long. [114/100]<br>Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 8]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
...
+    else
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, support_letter_attachment_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Line is too long. [138/100]<br>Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 9]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "false"
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Redundant `self` detected.

**Comment by RuslanHamidullin** [comment]
@meal just to confirm question :smile: 
did you test:
1) Upload of Form attachment's
2) Upload of Support Letters for Enterprise Promotion Form
3) Uploads of audit certificate on localhost:3000/users/form_answers//audit_certificate
4) And then make sure that Scan models are created if ENV["DISABLE_VIRUS_SCANNER"] == "true"
and that scan.clean condition works properly on admin side admin/form_answers/

Asking because, I still see in app/models/audit_certificate.rb :
```
after_save :virus_scan unless ENV["DISABLE_VIRUS_SCANNER"] == "true"
```

**Comment by RuslanHamidullin** [comment]
tbh: I don't think Audit Certificate upload would work with:
```
after_save :virus_scan unless ENV["DISABLE_VIRUS_SCANNER"] == "true"
```
as Scan is not creating on local / dev / demo servers in this case

**Comment by meal** [comment]
good catch, I forgot to remove it, fixed.

File: app/models/audit_certificate.rb [hunk 10]
```diff
@@ -33,7 +33,11 @@ def reviewed?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, audit_certificate_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Line is too long. [130/100]<br>Redundant `self` detected.

File: app/models/audit_certificate.rb [hunk 11]
```diff
@@ -33,7 +33,11 @@ def reviewed?
...
+    else
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
+      Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Line is too long. [106/100]<br>Redundant `self` detected.

File: app/models/audit_certificate.rb [hunk 12]
```diff
@@ -33,7 +33,11 @@ def reviewed?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], audit_certificate_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, audit_certificate_id: self.id, status: "clean")
+    else
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
```
Comment (matthewford) [inline app/models/audit_certificate.rb]: Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 13]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.file.url)
-    Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.file.current_path, uuid: SecureRandom.uuid, form_answer_attachment_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Line is too long. [129/100]<br>Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 14]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
...
+    else
+      scan = ::VirusScanner::File.scan_url(self.file.url)
+      Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Line is too long. [105/100]<br>Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 15]
```diff
@@ -29,7 +29,11 @@ def uploaded_not_by_user?
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.file.url)
-    Scan.create(filename: self.file.current_path, uuid: scan["id"], form_answer_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.file.current_path, uuid: SecureRandom.uuid, form_answer_attachment_id: self.id, status: "clean")
+    else
+      scan = ::VirusScanner::File.scan_url(self.file.url)
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 16]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, support_letter_attachment_id: self.id, status: "clean")
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Line is too long. [138/100]<br>Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 17]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
...
+    else
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
+      Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Line is too long. [114/100]<br>Redundant `self` detected.

File: app/models/support_letter_attachment.rb [hunk 18]
```diff
@@ -19,7 +19,11 @@ class SupportLetterAttachment < ActiveRecord::Base
   end
 
   def virus_scan
-    scan = ::VirusScanner::File.scan_url(self.attachment.url)
-    Scan.create(filename: self.attachment.current_path, uuid: scan["id"], support_letter_attachment_id: self.id)
+    if ENV["DISABLE_VIRUS_SCANNER"] == "true"
+      Scan.create(filename: self.attachment.current_path, uuid: SecureRandom.uuid, support_letter_attachment_id: self.id, status: "clean")
+    else
+      scan = ::VirusScanner::File.scan_url(self.attachment.url)
```
Comment (matthewford) [inline app/models/support_letter_attachment.rb]: Redundant `self` detected.

File: app/models/form_answer_attachment.rb [hunk 19]
```diff
@@ -29,7 +29,22 @@ def uploaded_not_by_user?
...
+      response = ::VirusScanner::File.scan_url(attachment.url)
+      Scan.create(
+        filename: attachment.current_path,
+        uuid: response["id"],
+        status: response["status"],
+        audit_certificate_id: id
+      )
+    end
   end
+
```
Comment (matthewford) [inline app/models/form_answer_attachment.rb]: Extra empty line detected at class body end.

**Comment by matthewford** [comment]
@RuslanHamidullin there is more work I want to do on this - but lets merge this as it is now. Note: DISABLE_VIRUS_SCANNER should be set to true while we work on this.

**Comment by RuslanHamidullin** [comment]
@matthewford thanks, merging

**Pull request merged**